com.unity.netcode.gameobjects/Runtime/SceneManagement/NetworkSceneManager.cs

```diff
@@ -8,101 +8,360 @@
         /// </summary>
-        public event SceneSwitchedDelegate OnSceneSwitched;
+        public ulong ClientId;
 
         /// <summary>
-        /// Event that is invoked when a local scene switch has started
+        /// List of clients that completed a loading or unloading event
+        /// Applies only to:
+        /// <see cref="SceneEventData.SceneEventTypes.S2C_LoadComplete"/>
+        /// <see cref="SceneEventData.SceneEventTypes.S2C_UnLoadComplete"/>
         /// </summary>
-        public event SceneSwitchStartedDelegate OnSceneSwitchStarted;
+        public List<ulong> ClientsThatCompleted;
 
         /// <summary>
-        /// Event that is invoked on the server when a client completes scene transition
+        /// List of clients that timed out during a loading or unloading event
+        /// Applies only to:
+        /// <see cref="SceneEventData.SceneEventTypes.S2C_LoadComplete"/>
+        /// <see cref="SceneEventData.SceneEventTypes.S2C_UnLoadComplete"/>
         /// </summary>
-        public event NotifyServerClientLoadedSceneDelegate OnNotifyServerClientLoadedScene;
+        public List<ulong> ClientsThatTimedOut;
+    }
+
+    /// <summary>
+    /// Main class for managing network scenes when <see cref="NetworkConfig.EnableSceneManagement"/> is enabled.
+    /// Uses the <see cref="MessageQueueContainer.MessageType.SceneEvent"/> message to communicate <see cref="SceneEventData"/> between the server and client(s)
+    /// </summary>
+    public class NetworkSceneManager
+    {
+        // Used to be able to turn re-synchronization off for future snapshot development purposes.
+        internal static bool DisableReSynchronization;
 
         /// <summary>
-        /// Event that is invoked on the server when all clients have reported that they have completed scene transition
+        /// Used to detect if a scene event is underway
+        /// Only 1 scene event can occur on the server at a time for now.
         /// </summary>
-        public event NotifyServerAllClientsLoadedSceneDelegate OnNotifyServerAllClientsLoadedScene;
+        private static bool s_IsSceneEventActive = false;
 
         /// <summary>
-        /// Event that is invoked on the clients after all clients have successfully completed scene transition or timed out.
-        /// <remarks>This event happens after <see cref="OnNotifyServerAllClientsLoadedScene"/> fires on the server and the <see cref="NetworkConstants.ALL_CLIENTS_LOADED_SCENE"/> message is sent to the clients.
-        /// It relies on MessageSender, which doesn't send events from the server to itself (which is the case for a Host client).</remarks>
+        /// For multi-instance unit tests, set this to true if you are use the <see cref="NetworkSceneManager"/>
         /// </summary>
-        public event NotifyClientAllClientsLoadedSceneDelegate OnNotifyClientAllClientsLoadedScene;
+        internal static bool IsTesting;
```

Comment:
I actually do not like seeing this at all. I remember using this check somewhere down there to skip test scenes from being considered for synchronization but I'd handle that case differently. For instance, I'd wrap that code-block with `#if UNITY_EDITOR` (and maybe something like `BUILD_TESTING` if there's one?) and then rely on some other flag somewhere else, ideally either a Unity API or our testing helper flag telling us "hey! you're actually running inside a test". seeing this flag here and being used by tests sounds weird to me.

Reply:
This could be renamed and is only specific to multiInstance unit tests.  I will investigate if there is a way to handle this in an more potentially automated way.

Reply:
@MFatihMAR 
I removed IsTesting and now handle this only when in UNITY_EDITOR and NetworkSceneManager detects whether or not it is running under an NUnit test context or not (i.e. IsTesting?)

Reply:
**Submitted a better solution:**  
Don't check for any kind of testing at all (whether automated or setting a property) and make adjustments to the problematic unit test and change a property in NetworkSceneManager from private to internal.